File: src/algorithms/sorting/comparison/strand-sort/sources/StrandSort.java
import java.util.ArrayList;
import java.util.List;

public class StrandSort {
    public static int[] strandSort(int[] inputArray) { // @step:initialize
        List<Integer> remainingList = new ArrayList<>(); // @step:initialize
        for (int value : inputArray) remainingList.add(value); // @step:initialize

        int arrayLength = remainingList.size(); // @step:initialize
        if (arrayLength <= 1) return inputArray.clone(); // @step:initialize

        List<Integer> outputList = new ArrayList<>(); // @step:initialize

        while (!remainingList.isEmpty()) {
            // Extract a strand: pick elements forming an ascending sequence
            List<Integer> strand = new ArrayList<>(); // @step:extract-strand
            List<Integer> leftover = new ArrayList<>(); // @step:extract-strand
            strand.add(remainingList.get(0)); // @step:extract-strand

            for (int scanIndex = 1; scanIndex < remainingList.size(); scanIndex++) { // @step:compare
                int currentValue = remainingList.get(scanIndex);
                if (currentValue >= strand.get(strand.size() - 1)) { // @step:compare
                    strand.add(currentValue); // @step:extract-strand
                } else {
                    leftover.add(currentValue); // @step:extract-strand
                }
            }

            outputList = mergeTwoSortedLists(outputList, strand); // @step:merge-strand
            remainingList = leftover; // @step:extract-strand
        }

        // @step:mark-sorted

        int[] resultArray = new int[outputList.size()]; // @step:complete
        for (int resultIndex = 0; resultIndex < outputList.size(); resultIndex++) {
            resultArray[resultIndex] = outputList.get(resultIndex);
        }
        return resultArray; // @step:complete
    }

    private static List<Integer> mergeTwoSortedLists(List<Integer> leftList, List<Integer> rightList) {
        List<Integer> merged = new ArrayList<>();
        int leftPointer = 0;
        int rightPointer = 0;

        while (leftPointer < leftList.size() && rightPointer < rightList.size()) {
            if (leftList.get(leftPointer) <= rightList.get(rightPointer)) {
                merged.add(leftList.get(leftPointer++));
            } else {
                merged.add(rightList.get(rightPointer++));
            }
        }

        while (leftPointer < leftList.size()) merged.add(leftList.get(leftPointer++));
        while (rightPointer < rightList.size()) merged.add(rightList.get(rightPointer++));

        return merged;
    }
}

File: src/algorithms/sorting/comparison/strand-sort/sources/strand-sort.py
def strand_sort(input_array: list[int]) -> list[int]:  # @step:initialize
    remaining_array = input_array.copy()  # @step:initialize
    array_length = len(remaining_array)  # @step:initialize

    if array_length <= 1:  # @step:initialize
        return remaining_array[:]  # @step:initialize

    output_array: list[int] = []  # @step:initialize

    while remaining_array:
        # Extract a strand: pick elements forming an ascending sequence
        strand: list[int] = [remaining_array[0]]  # @step:extract-strand
        leftover: list[int] = []  # @step:extract-strand

        for scan_index in range(1, len(remaining_array)):  # @step:compare
            if remaining_array[scan_index] >= strand[-1]:  # @step:compare
                strand.append(remaining_array[scan_index])  # @step:extract-strand
            else:
                leftover.append(remaining_array[scan_index])  # @step:extract-strand

        # Merge the strand into the output
        output_array = merge_two_sorted_arrays(output_array, strand)  # @step:merge-strand

        remaining_array = leftover  # @step:extract-strand

    # @step:mark-sorted

    return output_array  # @step:complete


def merge_two_sorted_arrays(left_array: list[int], right_array: list[int]) -> list[int]:
    merged: list[int] = []
    left_pointer = 0
    right_pointer = 0

    while left_pointer < len(left_array) and right_pointer < len(right_array):
        if left_array[left_pointer] <= right_array[right_pointer]:
            merged.append(left_array[left_pointer])
            left_pointer += 1
        else:
            merged.append(right_array[right_pointer])
            right_pointer += 1

    merged.extend(left_array[left_pointer:])
    merged.extend(right_array[right_pointer:])
    return merged
